keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1036 @@
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers fail to update some resources, since they fail validation.
+
+If the CRD Validation Ratcheting feature were enabled the following occurs:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers are allowed to update their resources, as long as all their changes are valid.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This KEP is scoped to CRDs, but if [KEP-3937](https://github.com/kubernetes/enhancements/pull/3938) is implemented then we should expect it to also affect native type validation.
+
+
+#### Schema-Changing Tests May Need Refactoring
+
+This change could cause some previously valid tests of invalid schema changes to fail. Those tests will have to be updated by changing to use create, or by
+changing the invalid fields during an update.
+
+#### Detection of Breaking Schema Changes is Different
+
+Some users might have a process to detect breaking schema changes by applying
+their CRD and poking an existing resource. This KEP would make that process
+no longer reliable.
+
+##### Mitigations
+###### Use CRD-Schema-Checker
+
+[CRD-Schema-Checker](https://github.com/openshift/crd-schema-checker) may be
+used to lint the CRD schema for breaking changes before applying it to a cluster.
+
+###### Use New Objects
+
+Users may cerate new objects rather than updating old ones to test for breaking schema changes. This feature has no effect on the creation code path.
+
+### Risks and Mitigations
+
+Help Wanted
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The basic idea is to add a post-processing step to the `Validate(Update)` methods of the CRD strategy. Given a list of `StatusError`, we check the field path associated with the error, if it is unchanged from old to new then the error is changed into a warning.
```

Comment:
I'm not super confident the field paths we get back from openapi validation are load-bearing and always resolveable jsonpaths.

In addition to questions about the reliability of the field paths, to do this post-processing, won't we have to:
1. iterate over every error
2. evaluate it as json-path on old and new
3. deep-compare the identified sections

We're already correlating pieces of the old object with pieces of the new object to accomodate CEL expressions using oldSelf. Did we consider re-using that logic to skip performance hits of step 2 (2x json path evaluations to find/correlate the invalid sections)?

Reply:
(also, would ratcheting logic be applied to cel validation expressions which don't use oldSelf?)

Reply:
> I'm not super confident the field paths we get back from openapi validation are load-bearing and always resolveable jsonpaths.

I'm not sure what you mean by load-bearing in this context, but yes the JSON paths become much more important with this change. A quick review seems to look like they are valid JSONPath strings, but a refactor to make the path tracking more robust (perhaps using the structured FieldPath rather than a path string) may be required.

> In addition to questions about the reliability of the field paths, to do this post-processing, won't we have to:

iterate over every error
evaluate it as json-path on old and new
deep-compare the identified sections

Yeah this accurately describes what I had in mind.

> We're already correlating pieces of the old object with pieces of the new object to accomodate CEL expressions using oldSelf. Did we consider re-using that logic to skip performance hits of step 2 (2x json path evaluations to find/correlate the invalid sections)?

CEL converts old and new object into a type that can resolve JSON paths easily. We definitely should look into sharing logic/instances. Ill add that to the KEP

> (also, would ratcheting logic be applied to cel validation expressions which don't use oldSelf?)

Yes

Reply:
> a refactor to make the path tracking more robust (perhaps using the structured FieldPath rather than a path string) may be required.

That would be... difficult... all the schema validation code lives in k8s.io/kube-openapi, and the translation from string names to field errors happens in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go#L52-L128

k8s.io/kube-openapi can't take a dependency on a staging repo like apimachinery

Reply:
I was expecting something more like the `obj, oldObj interface{}` correlation/threading in the CEL Validator#Validate method happening inside the SchemaValidator#Validate() method, with additional function signatures to pass in an old object and get back a distinct list of errors (either where the old and new object both failed, or where the old and new object stanzas were identical)

Reply:
> k8s.io/kube-openapi can't take a dependency on a staging repo like apimachinery
kube-openapi couldn't use field.Path directly, but it is a pretty simple type to derive

But yeah I think your suggestion of handling this logic while evaluating the errors in the first case is more direct/simpler. Thank you! Will look into that approach first!